instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java

```diff
@@ -0,0 +1,143 @@
+    ChannelPipeline channelPipeline = channel.pipeline();
+    HttpClientCodec handler = new HttpClientCodec();
+
+    // remove the default head and tail handlers
+    channelPipeline.removeFirst();
+    channelPipeline.removeLast();
+
+    // no handlers initially
+    assertEquals(0, channelPipeline.toMap().size());
+
+    // add handler
+    channelPipeline.addLast("http", handler);
+    assertEquals(handler, channelPipeline.first());
+    // our handler was also added
+    assertEquals("HttpClientTracingHandler", channelPipeline.last().getClass().getSimpleName());
+
+    if ("by instance".equals(testName)) {
+      channelPipeline.remove(handler);
+    } else if ("by class".equals(testName)) {
+      channelPipeline.remove(handler.getClass());
+    } else if ("by name".equals(testName)) {
+      channelPipeline.remove("http");
+    } else if ("first".equals(testName)) {
+      channelPipeline.removeFirst();
+    }
+
+    // removing handler also removes our handler
+    assertEquals(0, channelPipeline.toMap().size());
+  }
+
+  @ParameterizedTest
+  @CsvSource({"by instance", "by class", "by name"})
+  void shouldReplaceHandler(String desc) {
+    EmbeddedChannel channel = new EmbeddedChannel(new NoopChannelHandler());
+    ChannelPipeline channelPipeline = channel.pipeline();
+    HttpClientCodec httpHandler = new HttpClientCodec();
+
+    // remove the default head and tail handlers
+    channelPipeline.removeFirst();
+    channelPipeline.removeLast();
+
+    // no handlers initially
+    assertEquals(0, channelPipeline.toMap().size());
+
+    NoopChannelHandler noopHandler = new NoopChannelHandler();
+    channelPipeline.addFirst("test", noopHandler);
+
+    // only the noop handler
+    assertEquals(1, channelPipeline.toMap().size());
+    assertEquals(noopHandler, channelPipeline.first());
+
+    if ("by instance".equals(desc)) {
+      channelPipeline.replace(noopHandler, "http", httpHandler);
+    } else if ("by class".equals(desc)) {
+      channelPipeline.replace(noopHandler.getClass(), "http", httpHandler);
+    } else if ("by name".equals(desc)) {
+      channelPipeline.replace("test", "http", httpHandler);
+    }
+
+    // noop handler was removed; http and instrumentation handlers were added
+    assertEquals(2, channelPipeline.toMap().size());
+    assertEquals(httpHandler, channelPipeline.first());
+    assertEquals("HttpClientTracingHandler", channelPipeline.last().getClass().getSimpleName());
+
+    NoopChannelHandler anotherNoopHandler = new NoopChannelHandler();
+    channelPipeline.replace("http", "test", anotherNoopHandler);
+
+    // http and instrumentation handlers were removed; noop handler was added
+    assertEquals(1, channelPipeline.toMap().size());
+    assertEquals(anotherNoopHandler, channelPipeline.first());
+  }
+
+  // regression test for
+  // https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/4056
+  @Test
+  void shouldAddAfterAndRemoveLastHandler() {
```

Comment:
i modified this test based on current implementation. please let me know if this is not the case, might be regression? 
i might need to update netty 4.0 instrumentation for this. but i need some feedback here.